numpy/core/src/multiarray/lowlevel_strided_loops.c.src

```diff
@@ -1440,6 +1441,88 @@ PyArray_TransferMaskedStridedToNDim(npy_intp ndim,
 /****************** MapIter (Advanced indexing) Get/Set ********************/
 /***************************************************************************/
 
+/**begin repeat
+ * #name = checked, unchecked#
+ * #unchecked = 0, 1#
+ */
+
+/**begin repeat1
+ * #contig = , _contig#
+ * #iscontig = 0, 1#
+ * #indval_stride = _indval_stride, NPY_SIZEOF_INTP#
+ */
+
+NPY_NO_EXPORT int
+mapiter_update_pointers_@name@@contig@(
+        PyArrayMapIterObject *mit, npy_intp count, PyThreadState *_save)
+{
+    /* Specialize for a single iterator (could specialize function) */
```

Comment:
Ah, I think you need it.  There are two setups:
* The current one where we loop over each index array in one go.  In that case the first one is special because it adds the base pointer (not for optimization itself).
* We reverse the loop order.  In that case you have an unnecessary inner-for, that must certainly be slow.

I had changed it from the second to the first, so the comment isn't clear, because it only really fits to the second setup.

Reply:
Am not quite sure where your second point comes in, but I definitely had missed the `baseoffset` difference...
 
Still, I guess the question is, is it really much slower to change the loop to go from 0 to mit->numiter, and add
```
if (op == 0) {
    *pointer = baseoffset;
}
```
before l.1515 (`*pointers += curr_indval * fancy_stride;`)

I.e., the following (which I checked works, though I didn't do a performance check)

```
NPY_NO_EXPORT int
mapiter_update_pointers_@name@@contig@(
        PyArrayMapIterObject *mit, npy_intp count, PyThreadState *_save)
{
    char *baseoffset = mit->baseoffset;

    for (int op = 0; op < mit->numiter; op++) {
        npy_intp fancy_dim = mit->fancy_dims[op];
        npy_intp fancy_stride = mit->fancy_strides[op];
        char* indval = mit->outer_ptrs[op];
#if @unchecked@
        int iteraxis = mit->iteraxes[op];
#endif
#if !@iscontig@
        npy_intp _indval_stride = mit->outer_strides[op];
#endif
        char **pointers = mit->array_pointers;

        for (npy_intp i = 0; i < count; i++, indval += @indval_stride@, pointers++) {
            npy_intp curr_indval = *(npy_intp *)indval;
#if @unchecked@
            if (NPY_UNLIKELY(check_and_adjust_index(
                    &curr_indval, fancy_dim, iteraxis, _save) < 0 )) {
                return -1;
            }
#else
            if (curr_indval < 0) {
                curr_indval += fancy_dim;
            }
#endif
            if (op == 0) {
                *pointers = baseoffset;
            }
            *pointers += curr_indval * fancy_stride;
        }
    }

    return 0;
}
```

Reply:
That loses around 15-20% in the main `np.add.at()` trivial case (i.e. about the speedup you get for making a contiguous special case).

You can interchange the two loops:
```
for (npy_intp i = 0; i < count; i++, indval += @indval_stride@, pointers++) {
for (int op = 0; op < mit->numiter; op++) {
```
But doing so means you need to keep track of more pointers (swapping the loops may well be the real advantaged compared to the old setup, because you must swap them if you integrate the pointer calculation into the add/assign work loop).

One thing that shows how tricky perf judgement can be:  If I don't check index validity ahead of time, the speed is almost unchanged (~2% faster) on my Mac.

Reply:
Wow, that is surprising! But also worrying: it would mean that the trivial `ufunc.at` case depends really on little compiler optimizations. 

Overall, as it still seems worth keeping the trivial `ufunc.at` cases, perhaps this can be kept simple too?

Or, if not sure, how about splitting the PR in one that implements this for the general case, and another that can focus on seeing whether it can be made to work for the trivial `ufunc.at` loops too?

Reply:
To be fair, it depends. On my older intel laptop delaying the check turns out to be much faster.  Anyway, have to think about it.  The whole poitn this is faster, is that it is easier to optimize the simple loops.  I almost wonder if SIMDing the offset calculation would work (and if that is 2x faster, we beat the inlined version unless you SIMD that).

Reply:
You're right, that's a fused add-multiply, which should be quite fast...